File: src/types/transfer.ts
import { z } from 'zod';

export const transferProgressPayloadSchema = z.object({
  transferId: z.string(),
  connectionId: z.string(),
  fileName: z.string(),
  direction: z.enum(['upload', 'download']),
  bytesDone: z.number(),
  bytesTotal: z.number(),
  status: z.enum(['running', 'done', 'error']),
});

export type TransferProgressPayload = z.infer<typeof transferProgressPayloadSchema>;

export type TransferDirection = 'upload' | 'download';
export type TransferStatus = 'queued' | 'running' | 'done' | 'error' | 'skipped' | 'cancelled';

export interface TransferTask {
  id: string;
  connectionId: string;
  fileName: string;
  direction: TransferDirection;
  localPath: string;
  remotePath: string;
  isDirectory: boolean;
  status: TransferStatus;
  bytesDone: number;
  bytesTotal: number;
  error?: string;
}

export interface FileConflictInfo {
  fileName: string;
  localPath: string;
  remotePath: string;
  localSize: number;
  remoteSize: number;
  localModifiedAt?: string;
  remoteModifiedAt?: string;
}

File: src-tauri/src/utils/transfer.rs
use tauri::AppHandle;

use crate::events::emit_transfer_progress;

pub struct TransferProgress<'a> {
    app: &'a AppHandle,
    transfer_id: &'a str,
    connection_id: &'a str,
    file_name: &'a str,
    direction: &'a str,
    bytes_total: u64,
}

impl<'a> TransferProgress<'a> {
    pub fn new(
        app: &'a AppHandle,
        transfer_id: &'a str,
        connection_id: &'a str,
        file_name: &'a str,
        direction: &'a str,
        bytes_total: u64,
    ) -> Self {
        emit_transfer_progress(
            app,
            transfer_id,
            connection_id,
            file_name,
            direction,
            0,
            bytes_total,
            "running",
        );
        Self {
            app,
            transfer_id,
            connection_id,
            file_name,
            direction,
            bytes_total,
        }
    }

    pub fn update(&self, bytes_done: u64) {
        emit_transfer_progress(
            self.app,
            self.transfer_id,
            self.connection_id,
            self.file_name,
            self.direction,
            bytes_done.min(self.bytes_total),
            self.bytes_total,
            "running",
        );
    }

    pub fn done(&self) {
        emit_transfer_progress(
            self.app,
            self.transfer_id,
            self.connection_id,
            self.file_name,
            self.direction,
            self.bytes_total,
            self.bytes_total,
            "done",
        );
    }

    pub fn error(&self) {
        emit_transfer_progress(
            self.app,
            self.transfer_id,
            self.connection_id,
            self.file_name,
            self.direction,
            0,
            self.bytes_total,
            "error",
        );
    }
}
